src/trg-client.c

```diff
@@ -252,99 +257,72 @@ static GList *trg_client_headers_array_to_list(JsonArray *array)
         key = json_object_get_string_member(header_object, TRG_PREFS_KEY_CUSTOM_HEADER_NAME);
         value = json_object_get_string_member(header_object, TRG_PREFS_KEY_CUSTOM_HEADER_VALUE);
         if (key && value)
-            output_headers = g_list_prepend(output_headers, g_strdup_printf("%s: %s", key, value));
+            g_hash_table_insert(headers, (gpointer)key, (gpointer)value);
     }
-    return output_headers;
+
+    return headers;
 }
 
-int trg_client_populate_with_settings(TrgClient *tc)
+static void trg_client_inject_custom_header(gpointer key, gpointer value, gpointer user_data)
+{
+    SoupMessageHeaders *headers = user_data;
+    soup_message_headers_replace(headers, (gchar *)key, (gchar *)value);
+}
+
+gboolean trg_client_parse_settings(TrgClient *tc, gchar **err_msg)
 {
     TrgClientPrivate *priv = tc->priv;
     TrgPrefs *prefs = priv->prefs;
     JsonArray *headers;
-
+    g_autoptr(GError) uri_err = NULL;
     gint port;
-    gchar *host, *path;
-#if HAVE_LIBPROXY
-    pxProxyFactory *pf = NULL;
-#endif
 
     g_mutex_lock(&priv->configMutex);
 
     trg_prefs_set_connection(prefs, trg_prefs_get_profile(prefs));
 
-    g_clear_pointer(&priv->url, g_free);
+    g_clear_pointer(&priv->url, g_uri_unref);
     g_clear_pointer(&priv->username, g_free);
     g_clear_pointer(&priv->password, g_free);
-    g_list_free_full(priv->headers, g_free);
+    g_clear_pointer(&priv->headers, g_hash_table_unref);
     priv->headers = NULL;
 
     port = trg_prefs_get_int(prefs, TRG_PREFS_KEY_PORT, TRG_PREFS_CONNECTION);
-
-    host = trg_prefs_get_string(prefs, TRG_PREFS_KEY_HOSTNAME, TRG_PREFS_CONNECTION);
-    path = trg_prefs_get_string(prefs, TRG_PREFS_KEY_RPC_URL_PATH, TRG_PREFS_CONNECTION);
+    g_autofree gchar *host
+        = trg_prefs_get_string(prefs, TRG_PREFS_KEY_HOSTNAME, TRG_PREFS_CONNECTION);
+    g_autofree gchar *path
+        = trg_prefs_get_string(prefs, TRG_PREFS_KEY_RPC_URL_PATH, TRG_PREFS_CONNECTION);
 
     if (!host || strlen(host) < 1) {
-        g_free(host);
-        g_free(path);
         g_mutex_unlock(&priv->configMutex);
-        return TRG_NO_HOSTNAME_SET;
+        *err_msg = g_strdup("Bad hostname.");
+        return FALSE;
     }
 
     priv->ssl = trg_prefs_get_bool(prefs, TRG_PREFS_KEY_SSL, TRG_PREFS_CONNECTION);
     priv->ssl_validate
         = trg_prefs_get_bool(prefs, TRG_PREFS_KEY_SSL_VALIDATE, TRG_PREFS_CONNECTION);
 
-    priv->url = g_strdup_printf("%s://%s:%d%s", priv->ssl ? HTTPS_URI_PREFIX : HTTP_URI_PREFIX,
-                                host, port, path);
-    g_free(host);
-    g_free(path);
+    g_autofree gchar *uri_str = g_strdup_printf(
+        "%s://%s:%d%s", priv->ssl ? HTTPS_URI_PREFIX : HTTP_URI_PREFIX, host, port, path);
 
-    priv->username = trg_prefs_get_string(prefs, TRG_PREFS_KEY_USERNAME, TRG_PREFS_CONNECTION);
+    priv->url = g_uri_parse(uri_str, G_URI_FLAGS_NONE, &uri_err);
+    if (uri_err) {
+        g_mutex_unlock(&priv->configMutex);
```

Comment:
You might be interested in the [`GMutexLocker`](https://docs.gtk.org/glib/alias.MutexLocker.html) API. Nothing wrong here but I find it nicer to use.

Reply:
I think it should be possible to remove all the locking since we can restrict TrgPrefs to a single thread. That's for a later time though.